# VTK - ITK Project : Etude longitudinale de l'évolution d'une tumeur

### Par Raphaël Duhen, Maël Conan et Nigel Andrews

In [16]:
from typing import Tuple

import vtk
import itk

import matplotlib
import matplotlib.pyplot as plt
import numpy as np
import numpy.ma as ma

import ipywidgets as widgets
from ipywidgets import interact
from IPython.display import display, clear_output

In [17]:
def run_window(path: str):
    reader = vtk.vtkNrrdReader()
    reader.SetFileName(path)
    
    window = vtk.vtkRenderWindow()
    renderer = vtk.vtkRenderer()
    
    window.AddRenderer(renderer)
    
    interactor = vtk.vtkRenderWindowInteractor()
    window.SetInteractor(interactor)
    
    contour = vtk.vtkContourFilter()
    contour.SetInputConnection(reader.GetOutputPort())
    contour.SetValue(0, 135)
    
    contourMapper = vtk.vtkPolyDataMapper()
    contourMapper.SetInputConnection(contour.GetOutputPort())
    contourMapper.ScalarVisibilityOff()
    
    contourActor = vtk.vtkActor()
    contourActor.SetMapper(contourMapper)
    
    renderer.AddActor(contourActor)
    
    window.Render()
    interactor.Start()

In [18]:
PixelType = itk.D
Dimension = 3
ImageType = itk.Image[PixelType, Dimension]
TransformType = itk.TranslationTransform[PixelType, Dimension]

In [19]:
image = itk.imread("Data/case6_gre1.nrrd", pixel_type=PixelType)
image2 = itk.imread("Data/case6_gre2.nrrd", pixel_type=PixelType) 

In [20]:
def registrate(image, image2):
    dimension = image.GetImageDimension()
    TransformType = itk.TranslationTransform[itk.D, dimension]

    initial_transform = TransformType.New()
    initial_transform.SetIdentity()

    optimizer = itk.RegularStepGradientDescentOptimizerv4.New(
        LearningRate=4,
        MinimumStepLength=0.001,
        RelaxationFactor=0.5,
        NumberOfIterations=200,
    )

    metric = itk.MeanSquaresImageToImageMetricv4[ImageType, ImageType].New()

    registration = itk.ImageRegistrationMethodv4[ImageType, ImageType].New(
        Metric=metric,
        Optimizer=optimizer,
        InitialTransform=initial_transform,
        FixedImage=image,
        MovingImage=image2,
    )

    registration.Update()

    return registration, optimizer

In [21]:
def transform_image(image, image2, registration):
    transformed_image = itk.resample_image_filter(
        image2,
        transform=registration.GetTransform(),
        use_reference_image=True,
        reference_image=image,
        default_pixel_value=100,
    )

    return transformed_image

In [22]:
registration = None

try:
    transformed_image = itk.imread("Data/transformed_image.nrrd", pixel_type=PixelType)
except:
    registration, optimizer = registrate(image, image2)
    transformed_image = transform_image(image, image2, registration)

In [23]:
clear_output()

array_view = itk.array_view_from_image(transformed_image)
original_fixed_view = itk.array_view_from_image(image)
original_moving_view = itk.array_view_from_image(image2)

@interact(fixed=(0, original_fixed_view.shape[0]),
          moving=(0, original_moving_view.shape[0]),
          transformed=(0, array_view.shape[0]))
def plot_slices(fixed: int = 81, moving: int = 81, transformed: int = 81):
    plt.figure(figsize=(15, 15))

    plt.subplot(1, 3, 1)
    plt.imshow(original_fixed_view[fixed, :, :], cmap="gray", interpolation_stage='rgba')
    plt.title("Fixed image")

    plt.subplot(1, 3, 2)
    plt.imshow(original_moving_view[moving, :, :], cmap="gray", interpolation_stage='rgba')
    plt.title("Moving image")

    plt.subplot(1, 3, 3)
    plt.imshow(array_view[transformed, :, :], cmap="gray", interpolation_stage='rgba')
    plt.title("Transformed image")

    plt.show()

interactive(children=(IntSlider(value=81, description='fixed', max=176), IntSlider(value=81, description='movi…

In [24]:
itk.imwrite(transformed_image, "Data/transformed_image.nrrd")

In [25]:
if registration is not None:
    final_parameters = registration.GetOutput().Get().GetParameters()

    final_number_of_iterations = optimizer.GetCurrentIteration()

    print("Final parameters = ", final_parameters.GetElement(0))
    print("Final parameters = ", final_parameters.GetElement(1))
    print("Final parameters = ", final_parameters.GetElement(2))
    print("Final parameters = ", final_parameters.GetElement(3))
    print("Final parameters = ", final_parameters.GetElement(4))
    print("Final parameters = ", final_parameters.GetElement(5))

    print("Number of iterations = ", final_number_of_iterations)

In [26]:
def segmentation(image, lower: int = 400):
    PixelType = itk.D
    Dimension = 3
    ImageType = itk.Image[PixelType, Dimension]

    segmentation_filter = itk.ConnectedThresholdImageFilter[ImageType, ImageType].New(
        Lower=lower,
        Upper=1200,
        ReplaceValue=255,
        # This seed corresponds to a spot roughly near the center of the tumor.
        Seed=itk.Index[Dimension]([60, 120, 82]),
    )
    segmentation_filter.SetInput(image)
    segmentation_filter.Update()

    return segmentation_filter.GetOutput()

In [27]:
segmented_image = segmentation(transformed_image)

In [28]:
segmented_slices = itk.array_view_from_image(segmented_image)

In [29]:
masked_slices = ma.masked_array(array_view, mask=(segmented_slices == 0))

In [30]:
mask_cmap = matplotlib.cm.inferno
mask_cmap.set_bad('black', 1)

In [31]:
clear_output()

@interact(slice=(0, segmented_slices.shape[0] - 1))
def plot_segmentation(slice: int = 81):
    plt.figure(figsize=(15, 15))

    plt.subplot(1, 3, 1)
    plt.imshow(array_view[slice, :, :], cmap="inferno", interpolation_stage='rgba')
    plt.title("Transformed image")

    plt.subplot(1, 3, 2)
    plt.imshow(segmented_slices[slice, :, :], cmap="gray", interpolation_stage='rgba')
    plt.title("Segmented image")

    plt.subplot(1, 3, 3)
    plt.imshow(masked_slices[slice, :, :], cmap=mask_cmap, interpolation_stage='rgba')
    plt.title("Mask result")
    plt.show()

interactive(children=(IntSlider(value=81, description='slice', max=175), Output()), _dom_classes=('widget-inte…

In [32]:
box = [(0, segmented_slices.shape[0]), (0, segmented_slices.shape[1]), (0, segmented_slices.shape[2])]

@interact(
    x=widgets.IntRangeSlider(value=[103,141], min=0, max=segmented_slices.shape[1]),
    y=widgets.IntRangeSlider(value=[56,77], min=0, max=segmented_slices.shape[2]),
    z=widgets.IntRangeSlider(value=[69,89], min=0, max=segmented_slices.shape[0]),
    side_slice=(0, segmented_slices.shape[0]),
    top_slice=(0, segmented_slices.shape[1]),
)
def handfind_tumor_bounds(x: Tuple[int, int], y: Tuple[int, int], z: Tuple[int, int], side_slice: int = 81, top_slice: int = 66):
    plt.figure(figsize=(15, 15))

    plt.subplot(1, 2, 1)
    plt.imshow(masked_slices[side_slice, y[0]:y[1], x[0]:x[1]], cmap=mask_cmap, interpolation_stage='rgba')
    plt.title("Side view")

    ax = plt.subplot(1, 2, 2)
    plt.imshow(masked_slices[z[0]:z[1], top_slice, x[0]:x[1]], cmap=mask_cmap, interpolation_stage='rgba')
    plt.title("Top view")

    global box
    box = [tuple([*z]), tuple([*y]), tuple([*x])]

interactive(children=(IntRangeSlider(value=(103, 141), description='x', max=256), IntRangeSlider(value=(56, 77…

In [33]:
box_slice = [slice(*tup) for tup in box]
print(box_slice)

[slice(69, 89, None), slice(56, 77, None), slice(103, 141, None)]


In [34]:
box_mask = np.zeros(masked_slices.shape, dtype=bool)
box_mask[*box_slice] = True
boxed_tumor_slices = ma.masked_array(masked_slices, ~box_mask).filled(0)

In [35]:
clear_output()

@interact(slice=box[0])
def show_mask(slice: int = 81):
    plt.imshow(boxed_tumor_slices[slice, *box_slice[1:3]], cmap=mask_cmap)

interactive(children=(IntSlider(value=81, description='slice', max=89, min=69), Output()), _dom_classes=('widg…

In [36]:
# Uncomment to open tumor in viewer
# itk.imwrite(itk.image_view_from_array(boxed_tumor_slices), "Data/tumor2.nrrd")
# run_window("Data/tumor2.nrrd")

With the segmentation done for the second scan, we can replicate the process for the original scan, though we'll need to make use of morphological operations to clear the more proeminent noise.

In [51]:
def run():
    image = itk.imread("Data/case6_gre1.nrrd", pixel_type=itk.D)
    image_slices = itk.array_view_from_image(image)

    segmented = segmentation(image, lower=350)
    segmented_slices = itk.array_view_from_image(segmented) 

    masked_slices = ma.masked_array(image_slices, mask=(segmented_slices == 0))
    boxed_tumor_slices = ma.masked_array(masked_slices, ~box_mask).filled(0)

    return boxed_tumor_slices

original_slices = run()

@interact(slice=(70, 88))
def display(slice: int = 81):
    plt.imshow(original_slices[slice, *box_slice[1:3]], cmap=mask_cmap)

interactive(children=(IntSlider(value=81, description='slice', max=88, min=70), Output()), _dom_classes=('widg…

As we can see on slices 87 and 88, there is a large amount of noise. We'll get rid of these two slices as their positive impact on the volume of the original tumor will fake results.

In [53]:
original_slices = original_slices[:87]

In [54]:
# Uncomment to open original tumor in viewer
# itk.imwrite(itk.image_view_from_array(original_slices), "Data/tumor1.nrrd")
# run_window("Data/tumor1.nrrd")

# 4. Analyse et visualisation des changements

In [63]:
100 * np.count_nonzero(boxed_tumor_slices) / np.count_nonzero(original_slices) - 100

12.066856909906235

On remarque une augmentation du volume de la tumeur d'environ 12%.

In [62]:
100 * int(boxed_tumor_slices.sum(axis=2).sum()) / int(original_slices.sum(axis=2).sum()) - 100

9.486441507044262

L'intensité des voxels de la tumeur n'a quant à elle évolué que d'environ 9.5%, ce qui tend à indiquer que certaines parties de la tumeur ont une concentration en cellules cancéreuses moins élevée. Cette différence pourrait cependant aussi être attribuée à des erreurs d'approximation provenant des processus de recalage et de segmentation.